[Site Isolation] Cross-origin isolation should belong to the browsing context group

zakariaridouh opened the pull request:
#### cb6d451c2edf6babc126e8c4b27397338107a231
<pre>
[Site Isolation] Cross-origin isolation should belong to the browsing context group
<a href="https://bugs.webkit.org/show_bug.cgi?id=299201">https://bugs.webkit.org/show_bug.cgi?id=299201</a>
<a href="https://rdar.apple.com/116203642">rdar://116203642</a>

Reviewed by NOBODY (OOPS!).

A COOP: same-origin + COEP: require-corp response made a new browsing context group whose main
frame got a CrossOriginMode::Isolated process, but every other process the group used was
Shared: its cross-site subframes, the process picked after a crash, and the site isolation
shared process. A process cannot change its CrossOriginMode after launch, so the mode has to be
known before the process is created.

The mode is a capability, not a placement hint. WebProcessProxy::shouldEnableSharedArrayBuffer()
turns it into an XPC launch flag that sets JSC::Options::useSharedArrayBuffer() for the whole
process, and ScriptExecutionContext::setCrossOriginMode() raises the process-global timer
precision. So a Shared process denies SharedArrayBuffer to a subframe that is entitled to it,
and an Isolated process grants it to anything that lands there.

BrowsingContextGroup now holds the CrossOriginMode and every place that picks a process takes
it. An isolated group reuses nothing; processForNavigation() swaps to a fresh process whenever
the one it got does not match the group, covering every &quot;keep the current process&quot; exit in
processForNavigationInternal() except the two that promise never to swap.

Three things change because the mode now outlives a single navigation. The page adopts the new
group on commit rather than before it, so a download or a failed load cannot leave it marked
isolated - which means the group a navigation is in lives on the provisional page, and both
browsingContextGroupForNavigation() and the page creation parameters have to read it from
there. browsingContextGroupForNavigation() drops an isolated group for substitute data and
custom URL scheme handlers, the only main-frame navigations that never re-run COOP enforcement.
And a related page passed by a client no longer hands over an isolated group, though
createNewPage()&apos;s own uses of relatedPage still do: a blob: URL popup can only resolve its URL
in the opener&apos;s process, so denying it that process crashed the UI process.

Document::crossOriginIsolated() no longer goes through mainFrameDocument(), which is null when
the main frame is remote. Without this a cross-site subframe of a cross-origin-isolated page
reports self.crossOriginIsolated === false while having a SharedArrayBuffer, and agentClusterID()
keys it on site rather than origin, so two different origins on one site could share memory.

ServiceWorkerFetchTask::processResponse() now runs the navigation response&apos;s embedder policy
check. It validated CORP but not COEP inheritance, so a service worker could answer a subframe
navigation in a cross-origin-isolated group and be handed an isolated process without ever
opting in to COEP.

Unskips shared-array-buffers/window-failure.https.html under site isolation, which timed out
because its cross-site iframe could not make a SharedArrayBuffer.

Test: TestWebKitAPI.SiteIsolation.CrossOriginIsolated*
Test: TestWebKitAPI.SiteIsolation.ProcessesAreNotReusedAfterCrossOriginIsolatedBrowsingContextGroupSwitch
Test: TestWebKitAPI.SiteIsolation.RelatedWebViewDoesNotJoinCrossOriginIsolatedBrowsingContextGroup

* LayoutTests/platform/ios-site-isolation/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::crossOriginIsolated const): Resolve through the main frame&apos;s policy
container, which a RemoteFrame has, instead of mainFrameDocument().
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::doCrossOriginEmbedderPolicyHandlingOfNavigationResponse): Added,
for the response paths that do not go through didReceiveResponse().
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::processResponse): Enforce COEP inheritance.
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::preferredBrowsingContextGroup const): Only a client&apos;s related page is
refused an isolated group.
(API::PageConfiguration::crossOriginMode const): Added, so process selection and
getOrCreateBrowsingContextGroup() cannot disagree.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::setRelatedPage): Take a RelatedPageUse.
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPool.mm:
(-[WKProcessPool _crossOriginIsolatedProcessIdentifiersForTesting]): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPoolPrivate.h:
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::BrowsingContextGroup):
(WebKit::BrowsingContextGroup::sharedProcessForSite): Refuse the shared process when isolated.
(WebKit::BrowsingContextGroup::ensureProcessForSite): Check the mode where processes are bound
to groups, and don&apos;t fall into the shared process branch without a shared process.
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
(WebKit::BrowsingContextGroup::create):
(WebKit::BrowsingContextGroup::crossOriginMode const): Added.
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage): Pass the group this page will commit into.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::getOrCreateBrowsingContextGroup):
(WebKit::WebPageProxy::launchProcess): Use the group&apos;s mode, and compare the related page&apos;s
group rather than its process, which may be the dummy proxy.
(WebKit::WebPageProxy::suspendCurrentPageIfPossible): Keep the mode on the placeholder group.
(WebKit::WebPageProxy::browsingContextGroupForNavigation): Start from the provisional page&apos;s
group during a browsing context group switch, and drop an isolated group only for navigations
that skip COOP enforcement.
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision): Bind the frame&apos;s process to the
destination group, not the page&apos;s current one.
(WebKit::WebPageProxy::createNewPage): Load in the new process when it is not the opener&apos;s, and
don&apos;t build creation parameters for a page whose process has not launched.
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation): Tag the group, get the
process from processForSite() in that mode, and adopt the group on commit. Removed the FIXME.
(WebKit::WebPageProxy::triggerProcessSwapForEnhancedSecurity): Match the existing group&apos;s mode.
(WebKit::WebPageProxy::creationParameters): Take the destination browsing context group.
(WebKit::WebPageProxy::creationParametersForProvisionalPage): Ditto.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::tryTakePrewarmedProcess): Match the mode like lockdown mode and
enhanced security.
(WebKit::WebProcessPool::processForSite): Take a CrossOriginMode; an isolated group reuses
nothing.
(WebKit::WebProcessPool::createWebPage): Use the configuration&apos;s mode, and set
UsesSingleWebProcess here so every port reaches NetworkResourceLoader&apos;s single-process check.
(WebKit::WebProcessPool::crossOriginIsolatedProcessIdentifiers): Added.
(WebKit::WebProcessPool::processForNavigation): Check the mode on the back/forward and same-site
paths, and swap on a mismatch unless the client asked us never to.
(WebKit::WebProcessPool::prepareProcessForNavigation): Take the crash fallback&apos;s mode from the
process it replaces.
(WebKit::WebProcessPool::processForNavigationInternal): Swap on a source process mode mismatch,
alongside the other attributes fixed at launch.
* Source/WebKit/UIProcess/WebProcessPool.h:
* Tools/Scripts/webkitpy/api_tests/allowlist.txt:
* Tools/TestWebKitAPI/Tests/WebKit/WKWebView/SiteIsolation.mm:
(TestWebKitAPI::crossOriginIsolationServer): Added.
(TestWebKitAPI::isCrossOriginIsolatedProcess): Added.
(TestWebKitAPI::expectAllProcessesCrossOriginIsolated): Added.
(TestWebKitAPI::processIdentifiers): Added.
(TestWebKitAPI::expectSharedArrayBuffer): Added.
(TestWebKitAPI::expectCrossOriginIsolated): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupUsesIsolatedProcesses)): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupUsesIsolatedProcessesForNestedFrames)): Added.
(TestWebKitAPI::TEST(SiteIsolation, ProcessesAreNotReusedAfterCrossOriginIsolatedBrowsingContextGroupSwitch)): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupDoesNotUseSharedProcess)): Added.
(TestWebKitAPI::TEST(SiteIsolation, RelatedWebViewDoesNotJoinCrossOriginIsolatedBrowsingContextGroup)): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupWindowOpen)): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupNotAdoptedWhenNavigationIsCancelled)): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupAfterCrash)): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupAfterBackForwardNavigation)): Added.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb6d451c2edf6babc126e8c4b27397338107a231

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/183215 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/57138 "Hash cb6d451c for PR 51027 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/50955 "Hash cb6d451c for PR 51027 does not build (failure)") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/193433 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/147504 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/655663b6-c62f-4d9e-9dfa-2db08c2b0397) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/185078 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/58480 "Hash cb6d451c for PR 51027 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/56873 "Hash cb6d451c for PR 51027 does not build (failure)") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/143000 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/99308 "Exiting early after 60 failures. 60 tests run. 61 failures") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/74487894-d9a1-41dd-bf45-56707c5e4045) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/186170 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/155/builds/58480 "Hash cb6d451c for PR 51027 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/168/builds/50955 "Hash cb6d451c for PR 51027 does not build (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/123427 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b47283a0-8ce7-4948-ad94-a0323adafccc) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/182542 "Passed tests") | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/155/builds/58480 "Hash cb6d451c for PR 51027 does not build (failure)") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/156/builds/56873 "Hash cb6d451c for PR 51027 does not build (failure)") | | | 
| [❌ 🧪 jsc-x86-64](https://ews-build.webkit.org/#/builders/168/builds/50955 "Hash cb6d451c for PR 51027 does not build (failure)") | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/155/builds/58480 "Hash cb6d451c for PR 51027 does not build (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/168/builds/50955 "Hash cb6d451c for PR 51027 does not build (failure)") | [✅ 🛠 gtk3-gcc](https://ews-build.webkit.org/#/builders/284/builds/4607 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/49785 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/156/builds/56873 "Hash cb6d451c for PR 51027 does not build (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/195374 "Built successfully") | | 
| | [❌ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/56807 "Hash cb6d451c for PR 51027 does not build (failure)") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/168/builds/50955 "Hash cb6d451c for PR 51027 does not build (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/152495 "Passed tests") | | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/57512 "Hash cb6d451c for PR 51027 does not build (failure)") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/168/builds/50955 "Hash cb6d451c for PR 51027 does not build (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/152191 "Passed tests") | | 
| | [❌ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/160/builds/57512 "Hash cb6d451c for PR 51027 does not build (failure)") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/129782 "Built successfully") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/126091 "Build was cancelled. Recent messages:Printed configuration") | | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/56020 "Hash cb6d451c for PR 51027 does not build (failure)") | [❌ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/168/builds/50955 "Hash cb6d451c for PR 51027 does not build (failure)") | | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/55391 "Hash cb6d451c for PR 51027 does not build (failure)") | | | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/55629 "Hash cb6d451c for PR 51027 does not build (failure)") | | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/55458 "Hash cb6d451c for PR 51027 does not build (failure)") | | | | 
<!--EWS-Status-Bubble-End-->

zakariaridouh committed:
[Site Isolation] Page loads with certain headers should use their own process
https://bugs.webkit.org/show_bug.cgi?id=299201
rdar://116203642

Reviewed by NOBODY (OOPS!).

Processes shouldn't be reused after triggerBrowsingContextGroupSwitchForNavigation.
With site isolation, COOP, COEP, CORP headers should cause a new browsing context group,
not just a new main frame process.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, TriggerBrowsingContextGroupSwitchNewProcess)):

achristensen07 reviewed (changes requested):
The intent of that comment is not that we should always behave like we got ::NewIsolatedGroup whenever site isolation is on.  Rather, when we get ::NewIsolatedGroup and site isolation is on we should switch BrowsingContextGroup in addition to WebProcessProxy.  That should make it so that if we then make an iframe navigate to a site we already had a process for, it would not reuse that process either.

zakariaridouh committed:
[Site Isolation] Cross-origin isolation should belong to the browsing context group
https://bugs.webkit.org/show_bug.cgi?id=299201
rdar://116203642

Reviewed by NOBODY (OOPS!).

A COOP: same-origin + COEP: require-corp response made a new browsing context group whose main
frame got a CrossOriginMode::Isolated process, but every other process the group used was
Shared: its cross-site subframes, the process picked after a crash, and the site isolation
shared process. A process cannot change its CrossOriginMode after launch, so the mode has to be
known before the process is created.

The mode is a capability, not a placement hint. WebProcessProxy::shouldEnableSharedArrayBuffer()
turns it into an XPC launch flag that sets JSC::Options::useSharedArrayBuffer() for the whole
process, and ScriptExecutionContext::setCrossOriginMode() raises the process-global timer
precision. So a Shared process denies SharedArrayBuffer to a subframe that is entitled to it,
and an Isolated process grants it to anything that lands there.

BrowsingContextGroup now holds the CrossOriginMode and every place that picks a process takes
it. An isolated group reuses nothing; processForNavigation() swaps to a fresh process whenever
the one it got does not match the group, covering every "keep the current process" exit in
processForNavigationInternal() except the two that promise never to swap.

Three things change because the mode now outlives a single navigation. The page adopts the new
group on commit rather than before it, so a download or a failed load cannot leave it marked
isolated - which means the group a navigation is in lives on the provisional page, and both
browsingContextGroupForNavigation() and the page creation parameters have to read it from
there. browsingContextGroupForNavigation() drops an isolated group for substitute data and
custom URL scheme handlers, the only main-frame navigations that never re-run COOP enforcement.
And a related page passed by a client no longer hands over an isolated group, though
createNewPage()'s own uses of relatedPage still do: a blob: URL popup can only resolve its URL
in the opener's process, so denying it that process crashed the UI process.

Document::crossOriginIsolated() no longer goes through mainFrameDocument(), which is null when
the main frame is remote. Without this a cross-site subframe of a cross-origin-isolated page
reports self.crossOriginIsolated === false while having a SharedArrayBuffer, and agentClusterID()
keys it on site rather than origin, so two different origins on one site could share memory.

ServiceWorkerFetchTask::processResponse() now runs the navigation response's embedder policy
check. It validated CORP but not COEP inheritance, so a service worker could answer a subframe
navigation in a cross-origin-isolated group and be handed an isolated process without ever
opting in to COEP.

Unskips shared-array-buffers/window-failure.https.html under site isolation, which timed out
because its cross-site iframe could not make a SharedArrayBuffer.

Test: TestWebKitAPI.SiteIsolation.CrossOriginIsolated*
Test: TestWebKitAPI.SiteIsolation.ProcessesAreNotReusedAfterCrossOriginIsolatedBrowsingContextGroupSwitch
Test: TestWebKitAPI.SiteIsolation.RelatedWebViewDoesNotJoinCrossOriginIsolatedBrowsingContextGroup

* LayoutTests/platform/ios-site-isolation/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::crossOriginIsolated const): Resolve through the main frame's policy
container, which a RemoteFrame has, instead of mainFrameDocument().
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::doCrossOriginEmbedderPolicyHandlingOfNavigationResponse): Added,
for the response paths that do not go through didReceiveResponse().
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::processResponse): Enforce COEP inheritance.
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::preferredBrowsingContextGroup const): Only a client's related page is
refused an isolated group.
(API::PageConfiguration::crossOriginMode const): Added, so process selection and
getOrCreateBrowsingContextGroup() cannot disagree.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::setRelatedPage): Take a RelatedPageUse.
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPool.mm:
(-[WKProcessPool _crossOriginIsolatedProcessIdentifiersForTesting]): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPoolPrivate.h:
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::BrowsingContextGroup):
(WebKit::BrowsingContextGroup::sharedProcessForSite): Refuse the shared process when isolated.
(WebKit::BrowsingContextGroup::ensureProcessForSite): Check the mode where processes are bound
to groups, and don't fall into the shared process branch without a shared process.
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
(WebKit::BrowsingContextGroup::create):
(WebKit::BrowsingContextGroup::crossOriginMode const): Added.
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage): Pass the group this page will commit into.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::getOrCreateBrowsingContextGroup):
(WebKit::WebPageProxy::launchProcess): Use the group's mode, and compare the related page's
group rather than its process, which may be the dummy proxy.
(WebKit::WebPageProxy::suspendCurrentPageIfPossible): Keep the mode on the placeholder group.
(WebKit::WebPageProxy::browsingContextGroupForNavigation): Start from the provisional page's
group during a browsing context group switch, and drop an isolated group only for navigations
that skip COOP enforcement.
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision): Bind the frame's process to the
destination group, not the page's current one.
(WebKit::WebPageProxy::createNewPage): Load in the new process when it is not the opener's, and
don't build creation parameters for a page whose process has not launched.
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation): Tag the group, get the
process from processForSite() in that mode, and adopt the group on commit. Removed the FIXME.
(WebKit::WebPageProxy::triggerProcessSwapForEnhancedSecurity): Match the existing group's mode.
(WebKit::WebPageProxy::creationParameters): Take the destination browsing context group.
(WebKit::WebPageProxy::creationParametersForProvisionalPage): Ditto.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::tryTakePrewarmedProcess): Match the mode like lockdown mode and
enhanced security.
(WebKit::WebProcessPool::processForSite): Take a CrossOriginMode; an isolated group reuses
nothing.
(WebKit::WebProcessPool::createWebPage): Use the configuration's mode, and set
UsesSingleWebProcess here so every port reaches NetworkResourceLoader's single-process check.
(WebKit::WebProcessPool::crossOriginIsolatedProcessIdentifiers): Added.
(WebKit::WebProcessPool::processForNavigation): Check the mode on the back/forward and same-site
paths, and swap on a mismatch unless the client asked us never to.
(WebKit::WebProcessPool::prepareProcessForNavigation): Take the crash fallback's mode from the
process it replaces.
(WebKit::WebProcessPool::processForNavigationInternal): Swap on a source process mode mismatch,
alongside the other attributes fixed at launch.
* Source/WebKit/UIProcess/WebProcessPool.h:
* Tools/Scripts/webkitpy/api_tests/allowlist.txt:
* Tools/TestWebKitAPI/Tests/WebKit/WKWebView/SiteIsolation.mm:
(TestWebKitAPI::crossOriginIsolationServer): Added.
(TestWebKitAPI::isCrossOriginIsolatedProcess): Added.
(TestWebKitAPI::expectAllProcessesCrossOriginIsolated): Added.
(TestWebKitAPI::processIdentifiers): Added.
(TestWebKitAPI::expectSharedArrayBuffer): Added.
(TestWebKitAPI::expectCrossOriginIsolated): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupUsesIsolatedProcesses)): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupUsesIsolatedProcessesForNestedFrames)): Added.
(TestWebKitAPI::TEST(SiteIsolation, ProcessesAreNotReusedAfterCrossOriginIsolatedBrowsingContextGroupSwitch)): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupDoesNotUseSharedProcess)): Added.
(TestWebKitAPI::TEST(SiteIsolation, RelatedWebViewDoesNotJoinCrossOriginIsolatedBrowsingContextGroup)): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupWindowOpen)): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupNotAdoptedWhenNavigationIsCancelled)): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupAfterCrash)): Added.
(TestWebKitAPI::TEST(SiteIsolation, CrossOriginIsolatedBrowsingContextGroupAfterBackForwardNavigation)): Added.
